eclipse-theia/theia
# Support http proxies

#### What it does

Closes https://github.com/eclipse-theia/theia/issues/7797

Adds general proxy support to the whole application (backend+frontend) by using a `RequestService` which comes with automatic proxy support. Other services and downstream applications are encouraged to use that service whenever requests to non-origin or non-localhost web-services are performed.

Also patches the `http` and `https` node.js imports for the extension host. This enables extensions to perform requests through proxies as well.

For electron there's currently a large TODO: Automatic OS proxy detection doesn't work yet, since that requires access to the electron-main process (see also https://github.com/eclipse-theia/theia/pull/10698).

#### How to test

Generally, testing this PR requires setting up firewall rules which block all outgoing traffic and configuring a local proxy rule. I used [Fiddler Classic](https://www.telerik.com/fiddler/fiddler-classic) (Windows only?) which sets itself up at `http://localhost:8888`. I assume Mac and Linux users have a simpler time configuring such a proxy.

Also, these tests should preferably performed using the Electron version. The browser version will automatically use the OS proxy for stuff like fetching vsx search results, whereas the electron version will fail if the proxy is not configured.

1. Open the vsx registry and search for `java`. Nothing should appear if outbound traffic is blocked.
2. Search in the preferences for `proxy` and set the `Http: Proxy` preference to your local proxy url.
3. Confirm that re-searching displays results (frontend request) and allows to install the extension (backend request).
4. Install the [`proxy-test` extension](https://github.com/eclipse-theia/theia/files/8380275/proxy-test-0.0.1.zip).
5. Execute the `Hello World Request` command. It will send a request to `https://www.google.com` and display the error message/success code.
6. Change the `Http: Proxy support` preference. `override`, `fallback` and `on` should all provide the same success result, while `off` should display an error, since the extension proxy support has been turned off.

---

There are now proxy settings for fetching packages from ovsx registries using the `theia download:plugins` command.

0. Delete your `plugins` folder.
1. Executing `theia download:plugins` should result in an error, since the outgoing download requests are blocked
2. Adding the `--proxy-url <url>` parameter should enable successful download of these extensions.
 
#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
